mailer.rst

```diff
@@ -0,0 +1,653 @@
+            $message->from('fabien@symfony.com');
+        }
+
+        public static function getSubscribedEvents()
+        {
+            return [MessageEvent::class => 'onMessageSend'];
+        }
+    }
+
+.. _mailer-twig-integration:
+
+Twig: HTML & CSS
+----------------
+
+The Mime component integrates with the :doc:`Twig template engine </templating>`
+to provide advanced features such as CSS style inlining and support for HTML/CSS
+frameworks to create complex HTML email messages. First, make sure Twig is instaled:
+
+.. code-block:: terminal
+
+    $ composer require twig
+
+HTML Content
+~~~~~~~~~~~~
+
+To define the contents of your email with Twig, use the
+:class:`Symfony\\Bridge\\Twig\\Mime\\TemplatedEmail` class. This class extends
+the normal :class:`Symfony\\Component\\Mime\\Email` class but adds some new methods
+for Twig templates::
+
+    use Symfony\Bridge\Twig\Mime\TemplatedEmail;
+
+    $email = (new TemplatedEmail())
+        ->from('fabien@symfony.com')
+        ->to(new NamedAddress('ryan@symfonycasts.com', 'Ryan'))
+        ->subject('Thanks for signing up!')
+
+        // path of the Twig template to render
+        ->htmlTemplate('emails/signup.html.twig')
+
+        // pass variable (name => value) to the template
+        ->context([
+            'expiration_date' => new \DateTime('+7 days'),
+            'username' => 'foo',
+        ])
+    ;
+
+Then, create the template:
+
+.. code-block:: html+twig
+
+    {# templates/emails/signup.html.twig #}
+    <h1>Welcome {{ email.toName }}!</h1>
+
+    <p>
+        You signed up as {{ username }} the following email:
+    </p>
+    <p><code>{{ email.to[0].address }}</code></p>
+
+    <p>
+        <a href="#">Click here to activate your account</a>
+        (this link is valid until {{ expiration_date|date('F jS') }})
+    </p>
+
+
+The Twig template has access to any of the parameters passed in the ``context()``
+method of the ``TemplatedEmail`` class and also to a special variable called
+``email``, which is an instance of
+:class:`Symfony\\Bridge\\Twig\\Mime\\WrappedTemplatedEmail`.
+
+Text Content
+~~~~~~~~~~~~
+
+When the text content of a ``TemplatedEmail`` is not explicitly defined, mailer
+will generate it automatically by converting the HTML contents into text. If you
+have `league/html-to-markdown`_ installed in your application,
+it uses that to turn HTML into Markdown (so the text email has some visual appeal).
+Otherwise, it applies the :phpfunction:`strip_tags` PHP function to the original
+HTML contents.
+
+If you want to define the text content yourself, use the ``text()`` method
+explained in the previous sections or the ``textTemplate()`` method provided by
+the ``TemplatedEmail`` class:
+
+.. code-block:: diff
+
+    + use Symfony\Bridge\Twig\Mime\TemplatedEmail;
+
+    $email = (new TemplatedEmail())
+        // ...
+
+        ->htmlTemplate('emails/signup.html.twig')
+    +     ->textTemplate('emails/signup.txt.twig')
+        // ...
+    ;
```

Comment:
I believe correct diff templates should indent all lines with 2 spaces, except from the ones containing `+ `/`- `)

Reply:
I think what I have is correct. Looking at other parts of the docs, a diff code block (at least the way I've been creating them) involves making a normal code block (with normal indentation) and then adding `+ ` or `- ` (plus then space) on the changed lines. That indeed makes those lines look a little "funny" - as the changed lines don't line up with the unchanged lines inside the raw rst (but it renders correctly).